taichi/backends/opengl/opengl_api.cpp

```diff
@@ -152,161 +155,194 @@ struct GLProgram {
+    check_opengl_error("glUnmapBuffer");
   }
 
-  void *map(GLbitfield access = GL_READ_ONLY) const {
-    glBindBuffer(GL_SHADER_STORAGE_BUFFER, id_);
+  void flush_mapped_region(size_t offset, size_t size) {
+    glBindBuffer(type, id);
     check_opengl_error("glBindBuffer");
-    void *p = glMapBuffer(GL_SHADER_STORAGE_BUFFER, access);
-    check_opengl_error("glMapBuffer");
-    TI_ASSERT_INFO(p, "glMapBuffer returned NULL");
-    return p;
+    glFlushMappedBufferRange(type, offset, size);
+    check_opengl_error("glFlushMappedBufferRange");
   }
 
-  void unmap() const {
-    glBindBuffer(GL_SHADER_STORAGE_BUFFER, id_);
+  void read_back(void *buffer, size_t offset, size_t size) {
+    glBindBuffer(type, id);
     check_opengl_error("glBindBuffer");
-    glUnmapBuffer(GL_SHADER_STORAGE_BUFFER);
-    check_opengl_error("glUnmapBuffer");
+    glGetBufferSubData(type, offset, size, buffer);
+    check_opengl_error("glGetBufferSubData");
+  }
+
+  void read_back(void *buffer) {
+    read_back(buffer, 0, size);
   }
 };
 
-struct GLBuffer : GLSSBO {
-  GLBufId index;
-  void *base;
-  size_t size;
+struct GLBufferAllocator {
+ private:
+  static constexpr size_t max_free_resident_size = 64 << 20;
 
-  GLBuffer(GLBufId index, void *base, size_t size)
-      : index(index), base(base), size(size) {
-    bind_data(base, size);
-    bind_index((int)index);
-  }
+  struct BufferKey {
+    GLbitfield access;
+    size_t size;
 
-  GLBuffer(GLBufId index) : index(index), base(nullptr), size(0) {
-    bind_index((int)index);
-  }
+    struct Hash {
+      size_t operator()(const BufferKey &k) const {
+        return (size_t(k.access) << 32) ^ k.size;
+      }
+    };
 
-  void copy_forward() {
-    bind_data(base, size);
-  }
+    bool operator==(const BufferKey &k) const {
+      return k.access == access && k.size == size;
+    }
+  };
 
-  void rebind(void *new_base, size_t new_size) {
-    base = new_base;
-    size = new_size;
-    bind_data(base, size);
-  }
+  std::list<std::unique_ptr<GLBuffer>> buffers_storage_;
 
-  void copy_back() {
-    copy_back(base, size);
-  }
+  std::unordered_multimap<BufferKey, GLBuffer *, BufferKey::Hash>
+      buffers_mapping;
+  std::unordered_multimap<BufferKey, GLBuffer *, BufferKey::Hash> free_mapping;
 
-  void copy_back(void *ptr, size_t len) {
-    if (!len)
-      return;
-    void *mapped = this->map();
-    TI_ASSERT(mapped);
-    std::memcpy(ptr, mapped, len);
-    this->unmap();
+ public:
+  GLBufferAllocator() {
   }
-};
 
-struct GLBufferTable {
-  std::map<GLBufId, std::unique_ptr<GLBuffer>> bufs;
+  void new_frame() {
+    size_t free_resident_size = 0;
+    for (auto &pair : free_mapping) {
+      free_resident_size += pair.first.size;
+
+      if (free_resident_size > max_free_resident_size) {
+        GLBuffer *buf = pair.second;
+
+        // TI_WARN("Release {}", (void *)buf);
 
-  GLBuffer *get(GLBufId id) {
-    return bufs.at(id).get();
+        {
+          auto buffer_iter =
```

Comment:
nit: https://en.cppreference.com/w/cpp/algorithm/remove and https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom